File: algorithms/java/src/test/java/leetcode/LC338CountingBits.java
package leetcode.lc338_counting_bits;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertArrayEquals;

/**
 * https://leetcode.com/problems/counting-bits/
 * #easy
 */
public final class LC338CountingBits {
    public final class Solution {
        public int[] countBits(final int num) {
            int[] result = new int[num + 1];
            int threshold = 1;
            for (int i = 0; i < result.length; i++) {
                if ((threshold << 1) <= i) {
                    threshold <<= 1;
                }
                if (i == 0) {
                    result[0] = 0;
                } else {
                    result[i] = 1 + result[i - threshold];
                }
            }
            return result;
        }
    }

    @Test
    public void test_2() throws Exception {
        assertArrayEquals(new int[]{0, 1, 1}, new Solution().countBits(2));
    }

    @Test
    public void test_5() throws Exception {
        assertArrayEquals(new int[]{0, 1, 1, 2, 1, 2}, new Solution().countBits(5));
    }
}

File: algorithms/code/leetcode/lc338_counting_bits/lc338_counting_bits.py
# https://leetcode.com/problems/counting-bits/

import unittest


class Solution:
    def countBits(self, n: int) -> list[int]:
        result = [0] * (n + 1)
        threshold = 1
        for i, _ in enumerate(result):
            if threshold << 1 <= i:
                threshold <<= 1
            if i == 0:
                result[0] = 0
            else:
                result[i] = 1 + result[i - threshold]
        return result


class TestCode(unittest.TestCase):
    def test_2(self) -> None:
        assert Solution().countBits(2) == [0, 1, 1]

    def test_5(self) -> None:
        assert Solution().countBits(5) == [0, 1, 1, 2, 1, 2]
